### DISCUSSION TITLE
I'm curious, where/when do the clocks get initialized?

### DISCUSSION THREAD

**Discussion by LinusHeu** [earlephilhower/arduino-pico#1644] [Q&A] [answered]
Context: Due to https://github.com/earlephilhower/arduino-pico/discussions/415 and https://github.com/earlephilhower/arduino-pico/discussions/1544#discussioncomment-6721568 I got curious about clocks, and I'm trying to better understand how this core works under the hood.
I added some clock_get_hz() to the beginning of main() in https://github.com/earlephilhower/arduino-pico/blob/master/cores/rp2040/main.cpp#L79 and the clocks look already inizialized there.
![grafik](https://github.com/earlephilhower/arduino-pico/assets/118672629/ee7ea305-7e58-4d40-b4e7-9fa9526c82a3)
Does the OTA bootloader initialize them?
I found this file: https://github.com/earlephilhower/arduino-pico/blob/ca4637d14d4b9ad9df2a9b97e41427748a45c4f0/ota/ota_clocks.c but I can't find any place where __wrap_clocks_init or clocks_init() gets called. And modifying that file by adding `clock_stop(clk_peri); clock_stop(clk_adc);` to the end of `void __wrap_clocks_init(void)` didn't have any noticeable effect.

**Comment by earlephilhower** [comment, marked as answer]
The OTA does init the clocks.  The `__wrap*` fcns wrap functions in the base SDK (i.e. instead of building a new SDK blob, the linker just will call the __wrap versions).  This is because the SDK ones pull in a lot of extra code (floating point math, IIRC) which makes the OTA bootloader much larger than it needs to be.

If you change things in the `ota` dir, you need to rerun the `make-ota.sh` script to make the OTA blob.  OTW, your changes won't ecer be used.  You also need to make sure that the generated OTA blob still fits in the 12K allocated for it, of course.

**Reply by earlephilhower** [reply]
Also be aware that the main app will *also* re-init the clocks.  It's completely unaware of the OTA boot process.  So if you mess with clocks in OTA, you'll want to do the same in the main app.

**Reply by LinusHeu** [reply]
Thank you! Such linker sorcery is something I haven't even started to really understand or use. :D
